support GoogleDrive

tomtastic opened the pull request:
See https://github.com/pgaskin/NickelMenu/issues/143#issuecomment-2282194408

lorenzolewis commented:
Have tested this fork on my Kobo Clara Color and it's working like a charm. The only things that were a bit difficult were spotting that the config needed to be modified that's mentioned in the linked comment above.

pgaskin reviewed (changes requested):
Sorry about the delayed review.

Looks good overall, but can you update the documentation and rename the action to `gdrive` for consistency?

tomtastic commented:
Renamed, but haven't compiled to check it's all still working.
I've added the config documentation and a short blurb about what to put in the `Kobo eReader.conf`.

marcoms commented:
Appreciate the work on this! Is anything blocking merging atm?

viraj700 commented:
Hi, I can no longer download this. Could it be reposted or something please? It says expired

danifuuu commented:
@pgaskin can you review/merge and release this? is there anything left to be done? maybe I can help

tomtastic commented:
> Hi, I can no longer download this. Could it be reposted or something please? It says expired
```shell
$ git clone https://github.com/pgaskin/NickelMenu.git
$ cd NickelMenu
$ gh pr checkout 167
```

tomtastic commented:
There's a failing check for :
```
[action_cc.cc#L111](https://github.com/pgaskin/NickelMenu/pull/167/files#annotation_30307126848)
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.23.15505
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.24.15672
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.24.15676
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.25.15875
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.26.16704
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.28.17623
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.28.17820
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.28.17826
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.28.17925
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.28.18220
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.29.18730
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.30.18838
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.31.19086
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.32.19501
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.33.19608
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.33.19611
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.33.19759
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.34.20097
one of symbols [_ZN14MoreController11googleDriveEv] not found in libnickel.so.1.0.0@4.35.20400
```
Because it seems the Google Drive support only appeared around firmware version ~4.38 or thereabouts :
```shell
❯ find Kobo\ Libre\ H2O/Firmware -name "KoboRoot.tgz" | sort -n | while read; do echo -n "$REPLY -> "; tar xvfO "$REPLY" usr/local/Kobo/libnickel.so.1.0.0 | strings - | grep _ZN14MoreController11googleDriveEv;done
Kobo Libre H2O/Firmware/kobopatch_4.32.19501/out/KoboRoot.tgz -> x ./usr/local/Kobo/libnickel.so.1.0.0
Kobo Libre H2O/Firmware/kobopatch_4.33.19759/out/KoboRoot.tgz -> x ./usr/local/Kobo/libnickel.so.1.0.0
Kobo Libre H2O/Firmware/kobopatch_4.38.21908/out/KoboRoot.tgz -> x ./usr/local/Kobo/libnickel.so.1.0.0_ZN14MoreController11googleDriveEv
Kobo Libre H2O/Firmware/kobopatch_4.38.23171/out/KoboRoot.tgz -> x ./usr/local/Kobo/libnickel.so.1.0.0_ZN14MoreController11googleDriveEv
```

MKolman commented on src/action_cc.cc:
```diff
@@ -95,14 +101,42 @@ NM_ACTION_(nickel_open) {
             return nm_action_result_silent();
         }
 
+        if (!strcmp(arg1, "library") && !strcmp(arg2, "gdrive")) {
```
This entire block can be removed not that gdrive support is already added above

MKolman commented on res/doc:
```diff
@@ -124,6 +124,9 @@
 #                                          library:shelves           - Collections
 #                                          library:pocket            - Articles
 #                                          library:dropbox           - Dropbox (4.18.13737+)
+#                                          library:gdrive            - GoogleDrive (4.18.13737+) Also modify the '.kobo/Kobo/Kobo eReader.conf' with:
```
```suggestion
#                                          library:gdrive            - GoogleDrive (4.36.21095+) Also modify the '.kobo/Kobo/Kobo eReader.conf' with:
```

randombunny commented:
I tested it on Kobo Clara BW firmware version 4.43.23418.
There's a small error in action_cc.cc - "gdrive" should be "godrive". I had to change it to make it work.

randombunny commented on src/action_cc.cc:
```diff
@@ -75,14 +75,20 @@ NM_ACTION_(nickel_open) {
         NM_LOG("nickel_open: detected firmware >15505 (new nav tab bar), checking special cases");
 
-        if (!strcmp(arg1, "library") && !strcmp(arg2, "dropbox")) {
+        if (!strcmp(arg1, "library") && (!strcmp(arg2, "dropbox") || !strcmp(arg2, "gdrive"))) {
```
"gdrive" here must be "godrive" to work

tomtastic replied:
See the updated res/doc : https://github.com/pgaskin/NickelMenu/blob/7a06c5b45e429542b0c8acc2bd6688c37750e7e3/res/doc#L127

pgaskin reviewed (commented):
Sorry about the delay, looks good overall.

avichou commented:
![IMG_0743](https://github.com/user-attachments/assets/deb6c673-1c59-4e9e-aade-cbdf7e63e1ea)

unable to connect, it just keeps showing "loading". am I supposed to add the credentials beforehands as suggested in the other thread ? i'm not sure how to do it correctly
